### PR TITLE
[libminpack] Fix depends on PKGBUILD

### PR DESCRIPTION
**Pull request by luau-project**
## Description

Recently, I authored the PKGBUILD script for ```mingw-w64-libminpack```, and it was added to the repository ([#20643](https://github.com/msys2/MINGW-packages/pull/20643)) with valuable contributions from @MehdiChinoune. I am still learning the way how MSYS2 works, but it seems -- in my possibly mistaken view -- that ```depends``` field on PKGBUILD has the purpose to not break dependencies in the package tree.

On environments (```mingw64``` and ```ucrt64```) such that libminpack is built by gfortran, the programs dynamically linked to libminpack depend on a few runtime DLLs:
* **libgcc_s_seh-1.dll**
* **libgfortran-5.dll**
* **libminpack.dll** (library built from [MINGW-packages repository](https://github.com/msys2/MINGW-packages))
* **libquadmath-0.dll**
* **libwinpthread-1.dll**

> [!TIP]
> 
> In my brief investigation, on ```mingw64``` and ```ucrt64``` environments, I found that the dependency tree of ```${MINGW_PACKAGE_PREFIX}-fc-libs``` contains all the required runtime DLLs listed above.

> [!NOTE]
> 
> On ```clang64``` and ```clangarm64``` environments, there is no need to change anything because the app runs fine without any DLLs.

## TLDR;

In my view, for the environments ```mingw64``` and ```ucrt64```, a solution to handle dependencies properly would be to include the following command in the beginning of the PKGBUILD script.

```bash
depends=($([[ ${MSYSTEM} != CLANG* ]] && echo "${MINGW_PACKAGE_PREFIX}-fc-libs"))
```

In the following, I am going to show that an app dynamically linked to libminpack built on UCRT64 environment
asks for missing DLLs at runtime. First, I do an in-tree run of the app showing that everything works fine.
Later, I try to startup the same app out-of-MSYS2-tree and the system requires those DLLs to run the app. Thus,
this demonstrates the lack of proper dependencies, because at the moment PKGBUILD does not list any dependency for ```mingw64``` and ```ucrt64``` environments.

## Summary of the detailed explanation
* In-tree UCRT64 test
* Out-of-tree UCRT64 test

## In-tree UCRT64 test

1. Open a shell for ```ucrt64``` environemnt, then execute each command below

    ```bash
    # go to temp
    cd /tmp

    # create a clean folder to run tests
    rm -rf libminpack-chkder-test && \
    mkdir -p libminpack-chkder-test && \
    cd libminpack-chkder-test

    # clone MINGW-packages repo to build libminpack
    git clone https://github.com/msys2/MINGW-packages.git

    # change to libminpack dir
    cd MINGW-packages/mingw-w64-libminpack

    # build package (internet connection required)
    makepkg-mingw --cleanbuild --syncdeps --force --noconfirm

    # delete possible file there
    rm -f chkder-test.f

    # download test file from official minpack
    wget https://www.netlib.org/minpack/ex/file20 -O chkder-test.f

    # download data for that test
    wget https://www.netlib.org/minpack/ex/file23 -O chkder-test.dat

    # build the test program
    gfortran.exe -o chkder-test chkder-test.f \
    -L"pkg/mingw-w64-ucrt-x86_64-libminpack/ucrt64/lib" \
    -lminpack

    # pass the location of libminpack.dll in front of PATH
    # to run the compiled test program
    PATH=pkg/mingw-w64-ucrt-x86_64-libminpack/ucrt64/bin:$PATH \
    ./chkder-test.exe < chkder-test.dat
    ```
2. The expected result is that the test program ran normally;
3. Yet in the shell above, copy the generated test binary and data to a location out-of-MSYS2-tree (I am supposing that ```%USERPROFILE%\Downloads``` is out of tree).

    ```bash
    # create a clean folder to run the test out-of-tree
    rm -rf "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test && \
    mkdir -p "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test

    # copy the test program
    cp chkder-test.exe "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test

    # copy the test data
    cp chkder-test.dat "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test

    # sanity-check that the test program and test data were copied
    ls "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test
    ```

4. Leave that shell opened to ease out-of-tree tests below.

## Out-of-tree UCRT64 test

> [!IMPORTANT]
> 
> Check that ```C:\msys64\ucrt64\bin``` is **NOT** on your PATH system environment variable.

1. Open a vanilla cmd prompt and execute the test

    ```cmd
    REM change to directory created previously
    cd %USERPROFILE%\Downloads\libminpack-chkder-test

    REM execute the program without any DLLs in the folder
    chkder-test.exe < chkder-test.dat
    ```

2. Expected result (system asking for libgfortran-5.dll) 
Screenshot from 2024-04-17 14-16-13

4. In the ```ucrt64``` shell opened, copy libgfortran-5.dll to the out-of-tree directory
    
    ```bash
    # copy libgfortran-5.dll
    cp ${MINGW_PREFIX}/bin/libgfortran-5.dll "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test

    # sanity-check that it is there
    ls "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test
    ```
5. At cmd, rerun the test

    ```cmd
    chkder-test.exe < chkder-test.dat
    ```
6. Expected result (system asking for libgcc_s_seh-1.dll) 
Screenshot from 2024-04-17 14-20-39

7. On ```ucrt64``` shell, copy libgcc_s_seh-1.dll to out-of-tree dir
    
    ```bash
    # copy libgcc_s_seh-1.dll
    cp ${MINGW_PREFIX}/bin/libgcc_s_seh-1.dll "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test

    # sanity-check that it is there
    ls "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test
    ```
8. At cmd, rerun the test

    ```cmd
    chkder-test.exe < chkder-test.dat
    ```
9. Expected result (system asking for libminpack.dll) 
Screenshot from 2024-04-17 14-19-55

10. On ```ucrt64``` shell, copy libminpack.dll to out-of-tree dir
    
    ```bash
    # copy libminpack.dll
    cp pkg/mingw-w64-ucrt-x86_64-libminpack/ucrt64/bin/libminpack.dll "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test

    # sanity-check that it is there
    ls "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test
    ```
11. At cmd, rerun the test

    ```cmd
    chkder-test.exe < chkder-test.dat
    ```
12. Expected result (system asking for libwinpthread-1.dll) 
Screenshot from 2024-04-17 17-17-41

13. On ```ucrt64``` shell, copy libwinpthread-1.dll to out-of-tree dir
    
    ```bash
    # copy libwinpthread-1.dll
    cp ${MINGW_PREFIX}/bin/libwinpthread-1.dll "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test

    # sanity-check that it is there
    ls "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test
    ```
14. At cmd, rerun the test

    ```cmd
    chkder-test.exe < chkder-test.dat
    ```
15. Expected result (system asking for libquadmath-0.dll) 
Screenshot from 2024-04-17 14-25-20

16. On ```ucrt64``` shell, copy libquadmath-0.dll to out-of-tree dir
    
    ```bash
    # copy libquadmath-0.dll
    cp ${MINGW_PREFIX}/bin/libquadmath-0.dll "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test

    # sanity-check that it is there
    ls "$(cygpath.exe $USERPROFILE)"/Downloads/libminpack-chkder-test
    ```
17. At cmd, rerun the test

    ```cmd
    chkder-test.exe < chkder-test.dat
    ```
18. Finally, since all the required DLLs are on the app folder, the app runs fine.

### CODE REVIEW

**Comment by Biswa96** [comment]
According to the CI log, the libminpack.dll is not directly linked with fortran libraries. See https://github.com/msys2/MINGW-packages/actions/runs/8729470818/job/23951473195?pr=20657#step:11:718

**Comment by luau-project** [comment]
@Biswa96 

> According to the CI log, the libminpack.dll is not directly linked with fortran libraries. See https://github.com/msys2/MINGW-packages/actions/runs/8729470818/job/23951473195?pr=20657#step:11:718

> [!IMPORTANT]
> 
> Indeed. You are right. It was my fault.

For the sake of completeness in this thread, in case someone needs it in the future,
I used the following C code, which was mostly taken from [debian repository](https://salsa.debian.org/science-team/minpack/-/blob/master/debian/tchkder.c?ref_type=heads),
but small edited to load ```libminpack.dll``` dynamically and get a function pointer to ```chkder_```.

```C
/*      driver for chkder example. */

#include <stdio.h>
#include <math.h>
// #include <minpack.h>

#include "windows.h"

typedef void (*pchkder)( int *m, int *n, double *x, double *fvec, double *fjec,
	       int *ldfjac, double *xp, double *fvecp, int *mode,
	       double *err  );

void fcn(int *m, int *n, double *x, double *fvec,
	 double *fjac, int *ldfjac, int *iflag);

int main()
{
  HMODULE libminpack = LoadLibraryA(
    "libminpack.dll"
  );

  pchkder chkder_ = (pchkder)GetProcAddress(
    libminpack,
    "chkder_"
  );

  int i, m, n, ldfjac, mode;
  double x[3], fvec[15], fjac[15*3], xp[3], fvecp[15], 
    err[15];
  int one=1, two=2;

  m = 15;
  n = 3;

  /*      the following values should be suitable for */
  /*      checking the jacobian matrix. */

  x[1-1] = 9.2e-1;
  x[2-1] = 1.3e-1;
  x[3-1] = 5.4e-1;

  ldfjac = 15;

  mode = 1;
  chkder_(&m, &n, x, fvec, fjac, &ldfjac, xp, fvecp, &mode, err);
  mode = 2;
  fcn(&m, &n, x, fvec, fjac, &ldfjac, &one);
  fcn(&m, &n, x, fvec, fjac, &ldfjac, &two);
  fcn(&m, &n, xp, fvecp, fjac, &ldfjac, &one);
  chkder_(&m, &n, x, fvec, fjac, &ldfjac, xp, fvecp, &mode, err);

  for (i=1; i<=m; i++)
    {
      fvecp[i-1] = fvecp[i-1] - fvec[i-1];
    }
  printf("\n      fvec\n");  
  for (i=1; i<=m; i++) printf("%s%15.7g",i%3==1?"\n     ":"", fvec[i-1]);
  printf("\n      fvecp - fvec\n");  
  for (i=1; i<=m; i++) printf("%s%15.7g",i%3==1?"\n     ":"", fvecp[i-1]);
  printf("\n      err\n");  
  for (i=1; i<=m; i++) printf("%s%15.7g",i%3==1?"\n     ":"", err[i-1]);
  printf("\n");
  return 0;
}

void fcn(int *m, int *n, double *x, double *fvec,
	 double *fjac, int *ldfjac, int *iflag)
{
  /*      subroutine fcn for chkder example. */

  int i;
  double tmp1, tmp2, tmp3, tmp4;
  double y[15]={1.4e-1, 1.8e-1, 2.2e-1, 2.5e-1, 2.9e-1, 3.2e-1, 3.5e-1,
		3.9e-1, 3.7e-1, 5.8e-1, 7.3e-1, 9.6e-1, 1.34, 2.1, 4.39};

  
  if (*iflag != 2) 

    for (i=1; i<=15; i++)
      {
	tmp1 = i;
	tmp2 = 16 - i;
	tmp3 = tmp1;
	if (i > 8) tmp3 = tmp2;
	fvec[i-1] = y[i-1] - (x[1-1] + tmp1/(x[2-1]*tmp2 + x[3-1]*tmp3));
      }
  else
    {
      for (i = 1; i <= 15; i++)
	{
	  tmp1 = i;
	  tmp2 = 16 - i;
	  
	  /* error introduced into next statement for illustration. */
	  /* corrected statement should read    tmp3 = tmp1 . */
	  
	  tmp3 = tmp2;
	  if (i > 8) tmp3 = tmp2;
	  tmp4 = (x[2-1]*tmp2 + x[3-1]*tmp3); tmp4=tmp4*tmp4;
	  fjac[i-1+ *ldfjac*(1-1)] = -1.;
	  fjac[i-1+ *ldfjac*(2-1)] = tmp1*tmp2/tmp4;
	  fjac[i-1+ *ldfjac*(3-1)] = tmp1*tmp3/tmp4;
	}
    }
  return;
}
```

Placed ```libminpack.dll``` in the same folder and compiled this code with both GCC and latest MSVC,
and both generated binaries ran normally without ask for any DLL.

You can close this PR.

Sorry for the inconvenience.
